<!-- -*- mode: markdown; coding: utf-8; fill-column: 60; ispell-dictionary: "english" -*- -->

<meta charset="utf-8"/>
<meta name="viewport" content="width=device-width,initial-scale=1"/>
<link rel="stylesheet" href="style.css">


# EDAF75 - lab 2: Testing the database

As usual we have to tell Jupyter to allow SQL:

In [2]:
%load_ext sql

Check SQL Version

In [3]:
import sqlite3
ver = sqlite3.sqlite_version_info
print("SQLite Version {}.{}.{}".format(ver[0], ver[1], ver[2]))

SQLite Version 3.35.5


And then we import our movie database

In [4]:
%sql sqlite:///theaters.sqlite

__init__() got an unexpected keyword argument 'bind'
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])


We want to make sure that SQLite3 really checks our foreign
key constraints -- to do that, we run:

In [5]:
%%sql
PRAGMA foreign_keys=ON;

Environment variable $DATABASE_URL not set, and no connect string given.
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])


## SQL statements to review at the lab session

Write SQL code for the following tasks:

+ Show the names of all movies.

In [ ]:
%%sql
SELECT MovieTitle
FROM   movie

+ Show the performance dates for one of the movies.

In [ ]:
%%sql
SELECT    MovieTitle, PerformanceId, StartTime, PerformanceDate, TheaterName
FROM      movie
LEFT JOIN performance
USING     (IMDBKey)
WHERE     MovieTitle="The Terminator"

+ Show all data concerning performances at a given theatere
  on a given date.

In [ ]:
%%sql
SELECT      *
FROM        movie
RIGHT JOIN  performance
USING       (IMDBKey)
RIGHT JOIN  theater
USING       (TheaterName)
WHERE       PerformanceDate="2022-12-19" AND TheaterName="Paramount Theater"

+ List all customers

In [16]:
%%sql
SELECT *
FROM   customer

 * sqlite:///theaters.sqlite
Done.


[('test123', 'John Smith', 'password123'),
 ('brian22', 'Brian Svensson', 'SeCURePassword4$!'),
 ('alice2002', 'Alice Johnson', 'a53g4hj7ks5b6hb!#$#'),
 ('jacob1576', 'Jacob Krucinski', 'moVieAccount7428@')]

+ List all tickets

In [28]:
%%sql
SELECT *
FROM   ticket

 * sqlite:///theaters.sqlite
Done.


[('757be03a362d7f90e28490730e2c014c', 1, 'test123'),
 ('95ed6ec0fe370bb6bcac5456b78bc1fa', 2, 'test123'),
 ('347bba659a6744f3f25c4cf012fe548e', 3, 'test123'),
 ('b88887153900dbff9dd6f82eb75587d6', 3, 'jacob1576'),
 ('62390aff22b642c3ee04e2ee071acb6f', 3, 'jacob1576'),
 ('0cc22e9084151bd9c1452ec009d79709', 4, 'alice2002'),
 ('71624c294e4dae2afd6ed07772dad948', 5, 'jacob1576'),
 ('bf5adbd0569c1b7ae96a91c4fb7cab5c', 19, 'brian22'),
 ('1bd0c98ed6e4602fedee40d6107af85c', 12, 'test123'),
 ('c1c394d284b9835c1215f18aace932fb', 13, 'test123'),
 ('ab4949f77ac1800dbf827554a334c7e3', 14, 'jacob1576'),
 ('c213288a92d6b3b47f600bac5dabc297', 17, 'jacob1576'),
 ('98118375c330c87b7318909ac548b627', 15, 'alice2002'),
 ('e6a2dd68dbe8fe944820a2092254ea56', 16, 'jacob1576'),
 ('f6fc1978deec3c47e8e4d733e31264f9', 18, 'brian22'),
 ('9ffd4ae796f9d6d00fc4232474e526b0', 2, 'alice2002')]

## Things to try out yourselves (not reviewed at the lab session)

During lab 3 you'll write a REST service for the database
you just created, and for the server to work properly, the
things below should work.

If you have any questions about the material below, feel
free to ask them at the QA sessions (see Moodle for how to
sign up for them).


+ Create a new ticket to some performance (i.e., insert a
  new row in your table of tickets).

In [1]:
%%sql
INSERT
INTO        Ticket(Username, PerformanceId)
VALUES      ("alice2002", 2)
RETURNING   TicketId;

UsageError: Cell magic `%%sql` not found.


In [ ]:
# Regular SQL failing, try using sqlite3 library instead
db = sqlite3.connect("theaters.sqlite")
c = db.cursor()
c.execute("""INSERT INTO Ticket(Username, PerformanceId) VALUES ("alice2002", 2) RETURNING TicketId""")

found = c.fetchone()
print(found[0])
if not found:
    raise ValueError
else:
    db.commit()
    c.close() 

In [27]:
%%sql
-- Delete an entry from above
DELETE
FROM   Ticket
WHERE  TicketId="6da2243c675d3f4b6bb7f719b91704ae"

 * sqlite:///theaters.sqlite
1 rows affected.


[]

In newer versions of SQLite (since version 3.35, released
  in March 2021), and in
  [PostgreSQL](https://www.postgresql.org/docs/current/sql-insert.html),
  we can get any value generated during an insert using the
  `INSERT...-RETURNING` statement:

~~~{.sql}
INSERT
INTO       students
VALUES     ('Amy', 3.9, 1200)
RETURNING  s_id
~~~


which would return the generated `s_id` for the new
  student.

  If your SQLite version is older than 3.35, and you can't
  upgrade, you can instead use the following idea: each row
  in a SQLite3 table has a `rowid` attribute, it is a unique
  integer which essentially tells in which order the rows
  were inserted, and it's not displayed in queries unless we
  ask for it. SQLite3 also have a function,
  `last_insert_rowid()`, which returns the `rowid` of the
  last inserted row of a table, so we can see the `s_id` of
  the most recently inserted student with the following
  query:

~~~{.sql}
SELECT s_id
FROM   students
WHERE  rowid = last_insert_rowid();
~~~


Now, check what ticket number we got for the ticket we
  created above (it should be the same as the ticket id,
  which should be a `randomblob`):

In [31]:
%%sql
SELECT TicketId
FROM   Ticket
WHERE  rowid = last_insert_rowid();

 * sqlite:///theaters.sqlite
Done.


[]

+ Try to insert two movie theaters with the same name (this
  should fail).

In [ ]:
%%sql


+ Try to insert a performance where the theater doesn’t
  exist in the database (this should fail).

In [ ]:
%%sql


+ Create a ticket where either the user or the performance
  doesn’t exist (this should fail).

In [ ]:
%%sql
